keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md

```diff
@@ -0,0 +1,343 @@
+
+Scale down will be done a usual way (check stabilization window in the [Stabilization Window][] section below and the [Algorithm details][] in the official HPA documentation)
+
+[Stabilization Window]: #stabilization-window
+[Algorithm details]: https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#algorithm-details
+
+#### Story 2: Scale up as fast as possible, scale down very gradually
+
+This mode is essential when you don’t want to risk scaling down very quickly.
+
+Create an HPA with the constraints:
+
+- `constraints`:
+  - `scaleUp`:
+    - `percent = 900` (i.e. increase number of pods 10 times per minute is ok).
+  - `scaleDown`:
+    - `pods = 1`
+    - `periodSeconds = 600` (i.e., scale down one pod every 10 min)
+
+All other parameters are not specified (default values are used)
+
+If the application is started with 1 pod, it will scale up with the following number of pods:
+
+    1 -> 10 -> 100 -> 1000
+
+So, it can reach `maxReplicas` very fast.
+
+Scaling down will be by one pod each 10 min:
+
+    1000 -> 1000 -> 1000 -> … (7 more min) -> 999
+
+#### Story 3: Scale up very gradually, usual scale down process
+
+This mode is essential when you want to increase capacity, but you want it to be very pessimistic.
+
+Create an HPA with the constraints:
+
+- `constraints`:
+  - `scaleUp`:
+    - `pods = 1`    (i.e., increase only by one pod per minute)
+
+All other parameters are not specified (default values are used)
+
+If the application is started with 1 pod, it will scale up very gradually:
+
+    1 -> 2 -> 3 -> 4
+
+Scale down will be done a usual way (check stabilization window in [Algorithm details][])
+
+[Algorithm details]: https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#algorithm-details
+
+#### Story 4: Scale up as usual, do not scale down
+
+This mode is essential when you don’t want to risk scaling down at all.
+
+Create an HPA with the following constraints:
+
+- `constraints`:
+  - `scaleDown`:
+    - `percent = 0`
+    - `pods = 0`
+
+i.e., set both constraints to 0, so that the HPA controller would never scale the cluster down
+
+All other parameters are not specified (default values are used)
+
+The cluster will scale up as usual (default values), but will never scale down.
+
+### Implementation Details/Notes/Constraints
+
+#### Algorithm pseudocode
+
+The algorithm to find the number of pods will look like this:
+
+```golang
+desiredReplicas = 0
+calculatedReplicas = AnyAlgorithmInHPAController(...)
+if calculatedReplicas > curReplicas:
+  constraint1 = CurReplicas * (1 + ScaleUpPercent/100)
+  constraint2 = CurReplicas + ScaleUpPods
+  scaleUpLimit = max(constraint1, constraint2)
+  desiredReplicas = min(scaleUpLimit, calculatedReplicas)
+else if calculatedReplicas < CurReplicas:
+  constraint1 = curReplicas * (1 - ScaleDownPercent/100)
+  constraint2 = CurReplicas - ScaleDownPods
+  scaleDownLimit = min(constraint1, constraint2)
```

Comment:
I think of scale down limits as a mean to prevent too rapid a loss of capacity.  It would make more sense to me to apply `max(constraint1, constraint2)` during both scale-up and scale down.

Reply:
Agreed, that makes sense.

Reply:
Fixed: https://github.com/kubernetes/enhancements/pull/883/commits/8fe45e54c831378f35f24fc92c4957f905fd3254